velox/serializers/UnsafeRowSerde.cpp

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/serializers/UnsafeRowSerde.h"
+#include "velox/row/UnsafeRowDeserializer.h"
+#include "velox/row/UnsafeRowDynamicSerializer.h"
+
+namespace facebook::velox::serializer::batch {
+namespace {
+/// Finds row in a serializer vector by detecting the row (and key) sizes
+/// \param block block of bytes
+/// \param includesKeys true if the block has serialized keys as well
+/// in which case it is structured as follow: keySize keyBytes rowSize rowBytes
+/// \param rowPointers returned row pointers
+/// \return
+bool collectRowPointers(
+    const std::string_view& block,
+    bool includesKeys,
+    std::vector<std::optional<std::string_view>>& rowPointers) {
+  char* buffer = const_cast<char*>(block.data());
+  char* originalBuffer = buffer;
+
+  // Precess until the end of block
+  ssize_t remainingSize = block.size();
+  while (remainingSize > 0) {
+    // Skip keys if any
+    if (includesKeys) {
+      size_t keysSize = *((size_t*)buffer);
+      if (keysSize < 0 || keysSize > remainingSize) {
+        return false;
+      }
+      buffer += sizeof(keysSize);
+      buffer += keysSize;
+    }
+    // Read the row size
+    size_t rowSize = *((size_t*)buffer);
+    if (rowSize < 0 || rowSize > remainingSize) {
+      return false;
+    }
+    buffer += sizeof(rowSize);
+    rowPointers.emplace_back(std::string_view(buffer, rowSize));
+    buffer += rowSize;
+    remainingSize = block.size() - (buffer - originalBuffer);
+    if (remainingSize < 0) {
+      return false;
+    }
+  }
+  return true;
+}
+} // namespace
+
+std::optional<size_t> UnsafeRowVectorSerde::serializeKeys(
```

Comment:
The keys could be serialized by different serializer (with different format not necessary the unsaferow format) - i.e, Cosco is using it's own key serializer. It'd be better to make the key serializer pluggable.

Reply:
Very good idea! I moved it to a factory!

Reply:
@miaoever lmk if you have more comments on this PR